README.md
a simple lru kv cache library
==============================

Introduce
------------
A simple lru kv cache library,and implement similar as memcached lru.

Compile
----------
It is as simple as:
    
    % make

Test
--------
Using:

    % make test





ChangeLog
TODO:
	* add varint coding
	* support snappy and hugepage

2012-10-15
	* change readme


test/test.c
#include <stdio.h>
#include <assert.h>
#include "lru.h"

enum test_return { TEST_SKIP, TEST_PASS, TEST_FAIL };

typedef enum test_return (*test_func)(void);

struct testcase {
    const char *description;
    test_func function;
};

static enum test_return lru_normal_op(lru *l, size_t mb, char *key, size_t nkey, char *value, size_t nvalue) {
    char buf[64];
    size_t sz;
    int new = 0;

    if (!l) {
        l = lru_init(mb, 0);
        new = 1;
    }

    int r = item_set(l, key, nkey, value, nvalue);

    if (r) {
        if (new) {
            lru_free(l);
        }
        return TEST_FAIL;
    }

    r = item_get(l, key, nkey, buf, sizeof(buf), &sz);
    assert(r == 0);
    assert(sz == nvalue);
    assert(memcmp(value, buf, sz) == 0);

    r = item_delete(l, key, nkey);
    assert(r == 0);

    if (new) {
        lru_free(l);
    }
    return TEST_PASS;
}

static enum test_return lru_normal_test(void) {
    char *key = "key1";
    char *value = "value1";

    return lru_normal_op(NULL, 0, key, strlen(key), value, strlen(value) + 1);
}

static enum test_return lru_overwrite_test(void) {
    char *key = "key1";
    char *value = "value1";
    char *value2 = "value2";
    lru *l;

    l = lru_init(0, 0);

    item_set(l, key, strlen(key), value, strlen(value) + 1);

    enum test_return r = lru_normal_op(l, 0, key, strlen(key), value2, strlen(value2) + 1);

    lru_free(l);

    return r;
}

static enum test_return lru_big_set_test(void) {
    char *key = "key1";
    char *value = "value1";

    int malloc = ITEM_size + strlen(key) + 1 + strlen(value) + 1; 

    enum test_return r = lru_normal_op(NULL, malloc, key, strlen(key), value, strlen(value) + 1);

    if ( r != TEST_PASS) {
        return r;
    }

    r = lru_normal_op(NULL, malloc - 1, key, strlen(key), value, strlen(value) + 1);

    if ( r == TEST_FAIL) {
        return TEST_PASS;
    } else {
        return TEST_FAIL;
    }
}

static enum test_return lru_evict_test(void) {
    char key[5][5] = {"key1", "key2", "key3", "key4", "key5"};
    char *value = "value1";

    size_t nkey = 4;
    size_t nvalue = strlen(value);

    int malloc = ITEM_size + nkey + nvalue + 1; 
    int r, i = 0, loop;
    lru *l;

    loop = sizeof(key) / (nkey + 1);

    l = lru_init(malloc * loop, 0);

    size_t sz;
    char buf[256];


    for (i = 0; i < loop; i++) {
        r = item_set(l, key[i], nkey, value, nvalue);
        assert(r == 0);
    }

    for (i = 0; i < loop; i++) {
        r = item_get(l, key[i], nkey, buf, sizeof(buf), &sz);
        assert(r == 0);
        assert(sz == nvalue);
        assert(memcmp(value, buf, sz) == 0);
        memset(buf, 0, sizeof(buf));
    }

    //evict key1,key3 and replace key2
    char *key2 = "key2";
    size_t nvalue2 = malloc * 2 + 2;
    char value2[nvalue2];

    memset(value2, (int)'0', nvalue2);

    value2[nvalue2 -1] = '\0';
    value2[0] = 'a';

    r = item_set(l, key2, nkey, value2, nvalue2);
    assert(r == 0);

    for (i = 0; i < 3; i++) {
        if (i == 1) {
          continue;
        }
        r = item_get(l, key[i], nkey, buf, sizeof(buf), &sz);
        assert(r == 1);
    }
    for (i = 3; i < loop; i++) {
        r = item_get(l, key[i], nkey, buf, sizeof(buf), &sz);
        assert(r == 0);
        assert(sz == nvalue);
        assert(memcmp(value, buf, sz) == 0);
        memset(buf, 0, sizeof(buf));
    }

    r = item_get(l, key2, nkey, buf, sizeof(buf), &sz);
    assert(r == 0);
    assert(sz == nvalue2);
    assert(memcmp(value2, buf, sz) == 0);
    memset(buf, 0, sizeof(buf));

    lru_free(l);
    
    return TEST_PASS;
}

static enum test_return lru_stat_test(void) {
    char buf[1024];

    lru *l = lru_init(0, 0);
    char *key = "key1";
    char *value = "value1";

    lru_normal_op(l, 0, key, strlen(key), value, strlen(value) + 1);

    stat_print(l, buf, sizeof(buf));
    printf("%s\n", buf);

    stat_reset(l);

    stat_print(l, buf, sizeof(buf));
    printf("%s\n", buf);

    lru_free(l);

    return TEST_PASS;
}

struct testcase testcases[] = {
    { "normal_test", lru_normal_test },
    { "overwrite_test", lru_overwrite_test },
    { "big_test", lru_big_set_test },
    { "evict_test", lru_evict_test },
    { "stat_test", lru_stat_test },
    { NULL, NULL }
};

int main(void)
{
    int exitcode = 0;
    int ii = 0, num_cases = 0;

    for (num_cases = 0; testcases[num_cases].description; num_cases++);

    printf("1..%d\n", num_cases);

    for (ii = 0; testcases[ii].description != NULL; ++ii) {
        fflush(stdout);
        enum test_return ret = testcases[ii].function();
        if (ret == TEST_SKIP) {
            fprintf(stdout, "ok # SKIP %d - %s\n", ii + 1, testcases[ii].description);
        } else if (ret == TEST_PASS) {
            fprintf(stdout, "ok %d - %s\n", ii + 1, testcases[ii].description);
        } else {
            fprintf(stdout, "not ok %d - %s\n", ii + 1, testcases[ii].description);
            exitcode = 1;
        }
        fflush(stdout);
    }

    return exitcode;
}


test/bench.c
#include <stdlib.h>
#include <stdio.h>
#include <getopt.h>
#include <time.h>
#include <sys/time.h>
#include <assert.h>
#include <unistd.h>

#include "lru.h"

static struct config {
    size_t maxbytes;
    int num;
    int keysize;
    int datasize;
    int hashpower;
}config;

static void usage(void) {
    printf("-n <num>    total number of loop(default 1000,000)\n"
           "-m <num>    max memory for userd in megabytes(default 512MB)\n"
           "-k <num>    key size(default 16B)\n"
           "-d <num>    data size(default 32B)\n"
           "-p <num>    hash power level(default 16, 64K)\n"
           "-h          print this help and exit\n");
}

#define KEY_RAN_LEN 1024 * 1024 
static char *key_ = NULL;
static int pos_ = 0;
static int i_ = 0;
static int num_ = 0;
static char *fmt_;

static void generate_key_init() {
    char nbuf[64];
    num_ = snprintf(nbuf, sizeof(nbuf), "%d", config.num);

    snprintf(nbuf, sizeof(nbuf), "%%.*s%%0%dd",num_);
    fmt_ = strdup(nbuf);

    key_ = malloc(KEY_RAN_LEN);
    srandom(time(NULL));
    int i = 0;
    for ( i = 0; i < KEY_RAN_LEN; i++) {
        key_[i] = (char) (random() % 58) + 65; 
    }
}

static char* generate_key(int len) {
    if (pos_ + len > KEY_RAN_LEN) {
        pos_ = (++i_) % KEY_RAN_LEN;
    }
    pos_ += len;
    return key_ + pos_ - len;
}

static inline void generate_key_reset(void) {
    pos_ = 0;
    i_ = 0;
}

static void print_env(void) {
    printf("-----------------------------------------\n");
    printf("PID:         %d\n", (int)getpid());
    printf("ItemSize:    %zd\n", ITEM_size);
    printf("Keys:        %d bytes each\n", config.keysize);
    printf("Values:      %d bytes each\n", config.datasize);
    printf("Entries:     %d\n", config.num);

    printf("DataSize:    %.1f MB\n", (1L *(config.keysize + config.datasize) * config.num ) / 1048576.0);
    printf("TotalSize:   %.1f MB\n", (1L *(config.keysize + config.datasize + ITEM_size) * config.num ) / 1048576.0);
    printf("CacheSize:   %.1f MB\n", config.maxbytes / 1048576.0);
    printf("-----------------------------------------\n");
}

static long long start_;
static int done_, next_report_;
static char *benchmark_;

static long long mstime(void) {
    struct timeval tv;
    long long mst;

    gettimeofday(&tv, NULL);
    mst = ((long)tv.tv_sec)*1000;
    mst += tv.tv_usec/1000;
    return mst;
}


static void bench_start(char *bench) {
    start_ = mstime();
    benchmark_ = bench;
    done_ = 0;
    next_report_ = 0;
}

static void process_report(void) {
    done_++;
    if (done_ >= next_report_) {
        if (next_report_ < 1000) {
            next_report_ += 100;
        } else if (next_report_ < 5000) {
            next_report_ += 500;
        } else if (next_report_ < 10000) {
            next_report_ += 1000;
        } else if (next_report_ < 50000) {
            next_report_ += 5000;
        } else if (next_report_ < 100000) {
            next_report_ += 10000;
        } else if (next_report_ < 500000) {
            next_report_ += 50000;
        } else {
            next_report_ += 100000;
        }
        printf("... %s finished %d ops%30s\r", benchmark_, done_, "");
        fflush(stdout);
    }
}

static void bench_stop(void) {
    long long stop_ = mstime();
    printf("%-12s : take %lld ms, %11.3f op/s\n", benchmark_, (stop_ - start_),
         config.num * 1000.0 / (stop_ - start_));
}

static void config_init(void) {
    config.num = 1000000;
    config.maxbytes = 512 * 1024 * 1024;
    config.keysize = 16;
    config.datasize = 32;
    config.hashpower = 16;
}

static void print_stat(lru *l) {
    char bstat[1024];
    stat_print(l, bstat, sizeof(bstat));
    printf("-----------------------------------------\n");
    printf("STAT\n%s", bstat);
    printf("-----------------------------------------\n");
}

int main(int argc, char **argv) {
    config_init();

    int c;
    
    while (-1 != (c = getopt(argc, argv, "n:m:k:d:p:h"))) {
        switch(c) {
            case 'n':
                config.num = atoi(optarg);
                break;
            case 'm':
                config.maxbytes = ((size_t)atoi(optarg)) * 1024 * 1024;
                break;
            case 'k':
                config.keysize = atoi(optarg);
                break;
            case 'd':
                config.datasize = atoi(optarg);
                break;
            case 'p':
                config.hashpower = atoi(optarg);
                break;
            case 'h':
                usage();
                return EXIT_SUCCESS;
            default:
                usage();
                return EXIT_FAILURE;
        }
    }

    generate_key_init();
    print_env();
    lru *l = lru_init(config.maxbytes, config.hashpower);

    char *bvalue = malloc(config.datasize);
    memset(bvalue, 'x', config.datasize);

    char *key = malloc(config.keysize);
    memset(key, 0, config.keysize);

    int gnum = config.keysize - num_;

    generate_key_reset();
    bench_start("SET");
    int i;
    for (i = 0; i < config.num; i++) {
        snprintf(key, config.keysize, fmt_, gnum, generate_key(gnum), i);
        int r = item_set(l, key, config.keysize, bvalue, config.datasize);
        assert(r == 0);
        process_report();
    }
    bench_stop();
    print_stat(l);

    char *buf = malloc(config.datasize);
    size_t sz;
    generate_key_reset();
    bench_start("GET");
    for (i = 0; i < config.num; i++) {
        snprintf(key, config.keysize, fmt_, gnum, generate_key(gnum), i);
        int r = item_get(l, key, config.keysize, buf, config.datasize, &sz);
        if (!r) {
            assert((int)sz == config.datasize);
            assert(memcmp(bvalue, buf, config.datasize) == 0);
        }
        memset(buf, 0, config.datasize);
        process_report();
    }
    bench_stop();
    print_stat(l);

    generate_key_reset();
    bench_start("DELETE");
    for (i = 0; i < config.num; i++) {
        snprintf(key, config.keysize, fmt_, gnum, generate_key(gnum), i);
        item_delete(l, key, config.keysize);
        process_report();
    }
    bench_stop();
    print_stat(l);
    
    free(buf);
    free(bvalue);
    free(key);
    free(fmt_);
    free(key_);
   
    lru_free(l);
    /*
    printf("print any key to exit...\n");
    getchar();
    */
    return EXIT_SUCCESS;
}


lru.h
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#define MAX(a, b) (((a) > (b))?(a) : (b))
#define MIN(a, b) (((a) > (b))?(b) : (a))
#define MAXBYTE_DEDAULT 64 * 1024 * 1024 /* 64M */
#define HASH_POWER_DEFAULT 16

#define hashsize(n) ((size_t)1<<(n))
#define hashmask(n) (hashsize(n)-1)


#define ITEM_size   (sizeof(lru_item))

#define ITEM_key(item) ((char*)((item)->data))    /* terminal null */

#define ITEM_data(item) (char*) ((item)->data) + (item)->nkey + 1 

/* item */
typedef struct lru_item_ {
    struct lru_item_ *next;
    struct lru_item_ *prev;
    struct lru_item_ *h_next;   /* hash chain next */
    int nbytes;         /* size of total items, ITEM_size + nkey + 1 + nvalue*/
    uint8_t nkey;       /* key len, not include terminal null */
    char data[];        /* key and value, include terminal null */
}lru_item;


/* stat */
typedef struct lru_stat {
    uint64_t total_items;
    uint64_t curr_items;

    uint64_t total_bytes;
    uint64_t malloc;
    uint64_t malloc_failed;

    uint64_t free_bytes;
    uint64_t free;
    uint64_t curr_bytes;
    
    uint64_t get_cmds;
    uint64_t get_hits;
    uint64_t get_misses;

    uint64_t set_cmds;
    uint64_t set_failed;

    uint64_t del_cmds;
    uint64_t del_hits;
    uint64_t del_misses;

    unsigned int hash_find_depth;
    uint64_t hash_bytes;
    

    uint64_t evictions;
}lru_stat;

typedef struct lru{
    unsigned int hashpower;
    uint64_t max_bytes;  /* max used bytes, include item size, not include hash bytes */

    lru_item **table;   /* hash */
    lru_item *head;     /* double link head */
    lru_item *tail;     /* double link tail */

    lru_stat stat;      /* stat */
}lru;


#include "hash.h"

lru* lru_init(const uint64_t maxbytes, const unsigned int hashpower);

/*
   0 success 
   1 failed 
   alloc value buf by caller or add refcount by item...
*/
int item_get(lru *l, const char *key, const size_t nkey, char *buf, const size_t nbuf, size_t *nvalue);

/* 0 success , 1 failed */
int item_set(lru *l, const char *key, const size_t nkey, const char *value, const size_t nvalue);

/* 0 hit, 1 miss */
int item_delete(lru *l, const char *key, const size_t nkey);

#define stat_reset(l) do { \
    memset(&(l->stat), 0, sizeof(struct lru_stat)); \
} while(0)

void stat_print(lru *l, char *buf, const int nbuf);

void lru_free(lru *l);


hash.c
#include "lru.h"

lru_item*
hash_find(lru *l, const char *key, const size_t nkey, const uint32_t hv)
{
    lru_item *it = l->table[hv & hashmask(l->hashpower)];
    lru_item *ret = NULL;
    unsigned int depth = 0;
    while (it) {
        if ((nkey == it->nkey) && (memcmp(key, ITEM_key(it), nkey) == 0)) {
            ret = it;
            break;
        }
        it = it->h_next;
        ++depth;  
    }
    l->stat.hash_find_depth = MAX(l->stat.hash_find_depth, depth);
    return ret;
}

int 
hash_insert(lru *l, lru_item *it, const uint32_t hv)
{
    it->h_next = l->table[hv & hashmask(l->hashpower)];
    l->table[hv & hashmask(l->hashpower)] = it;
    return 1;
}

static lru_item** 
hashitem_before (lru *l, const char *key, const size_t nkey, const uint32_t hv) 
{
    lru_item **pos;

    pos = &(l->table[hv & hashmask(l->hashpower)]);

    while (*pos && ((nkey != (*pos)->nkey) || memcmp(key, ITEM_key(*pos), nkey))) {
        pos = &(*pos)->h_next;
    }
    return pos;
}

void 
hash_delete(lru *l, const char *key, const size_t nkey, const uint32_t hv)
{
    lru_item **before = hashitem_before(l, key, nkey, hv);

    if (*before) {
        lru_item *nxt;
        nxt = (*before)->h_next;
        (*before)->h_next = 0;  
        *before = nxt;
    }
}


binding/lru_jni.c
#include "lru_jni.h"
#include "lru.h"

/* open */
JNIEXPORT jlong JNICALL 
Java_com_feinno_kv_MemDB_open(JNIEnv *jenv, jobject clazz, jlong bufsize, jint hashpower)
{
    (void) clazz;
    (void) jenv;

    lru *l = lru_init((uint64_t) bufsize, (unsigned int) hashpower);
    return (jlong)(intptr_t)l;
}

/* set */
JNIEXPORT jint JNICALL 
Java_com_feinno_kv_MemDB_set(JNIEnv *jenv, jobject clazz, jlong ptr, jbyteArray jkey, jint jklen, jbyteArray jval, jint jvlen)
{
	(void)clazz;

	char *key, *val;
    int ret = -1;
    lru *l = (lru*) ptr;

	key = (char*)(*jenv)->GetByteArrayElements(jenv, jkey, 0);
	val = (char*)(*jenv)->GetByteArrayElements(jenv, jval, 0);

	if (key == NULL || val == NULL)
		goto RET;

	ret = item_set(l, key, (int)jklen, val, (int)jvlen);

RET:
	if (key) {
		(*jenv)->ReleaseByteArrayElements(jenv, jkey, (jbyte*)key, 0);
	}
	if (val){
		(*jenv)->ReleaseByteArrayElements(jenv, jval, (jbyte*)val, 0);
	}
	return ret;
}

/* get */
JNIEXPORT jint JNICALL 
Java_com_feinno_kv_MemDB_get(JNIEnv *jenv, jobject clazz, jlong ptr, jbyteArray jkey, jint jklen, jbyteArray jbuf, jint jblen)
{
    (void) clazz;
	char *key, *buf;
    int ret = -1;
    lru *l = (lru*) ptr;

	key = (char*)(*jenv)->GetByteArrayElements(jenv, jkey, 0);
	buf = (char*)(*jenv)->GetByteArrayElements(jenv, jbuf, 0);

	if (key == NULL)
		goto RET;

    size_t sz;
	ret = item_get(l, key, (int)jklen, buf, (int)jblen, &sz);

    //(*jenv)->SetByteArrayRegion(jenv, jbuf, 0, sz, (jbyte*)buf);
RET:
	if (key) {
		(*jenv)->ReleaseByteArrayElements(jenv, jkey, (jbyte*)key, 0);
	}
	if (buf) {
		(*jenv)->ReleaseByteArrayElements(jenv, jbuf, (jbyte*)buf, 0);
	}

	return ret ? 0 : sz;
}

/* remove */
JNIEXPORT jint JNICALL 
Java_com_feinno_kv_MemDB_remove(JNIEnv *jenv, jobject clazz, jlong ptr, jbyteArray jkey, jint jklen)
{
    (void) clazz;
	char *key;
    int ret = -1;
    lru *l = (lru*) ptr;

	key = (char*)(*jenv)->GetByteArrayElements(jenv, jkey, 0);

	if (key == NULL)
		goto RET;

	ret = item_delete(l, key, (int)jklen);

RET:
	if (key) {
		(*jenv)->ReleaseByteArrayElements(jenv, jkey, (jbyte*)key, 0);
	}
	return ret;
}


/* stat */
JNIEXPORT void JNICALL 
Java_com_feinno_kv_MemDB_info(JNIEnv *jenv, jobject clazz, jlong ptr, jbyteArray jbuf, jint jblen)
{
    (void) clazz;
    char *buf;
    lru *l = (lru*) ptr;

    buf = (char*)(*jenv)->GetByteArrayElements(jenv, jbuf, 0);
    if (buf) {
        stat_print(l, buf, (int)jblen);
		(*jenv)->ReleaseByteArrayElements(jenv, jbuf, (jbyte*)buf, 0);
    }
}

/* free */
JNIEXPORT void JNICALL 
Java_com_feinno_kv_MemDB_close(JNIEnv *jenv, jobject clazz, jlong ptr)
{
    (void) jenv;
    (void) clazz;
    lru *l = (lru*) ptr;

    lru_free(l);
}



lru.c
#include "lru.h"

#include <assert.h>
#include <stdarg.h>
#include <stdio.h>

lru* 
lru_init(const uint64_t maxbytes, const unsigned int hashpower)
{
    lru *l = calloc(sizeof(lru), 1);
    
    if (l == NULL) {
        fprintf(stderr, "lru init alloc lru failed.\n");
        return NULL;
    }

    if (hashpower > 32) {
        fprintf(stderr, "hash power must less than 32.\n");
        return NULL;
    } else {
        l->hashpower = hashpower == 0 ? HASH_POWER_DEFAULT: hashpower;
    }

    l->max_bytes = maxbytes ==0 ? MAXBYTE_DEDAULT: maxbytes;
    l->table = calloc(hashsize(l->hashpower), sizeof(void *));

    if (l->table == NULL) {
        fprintf(stderr, "lru init alloc hash buckets failed.\n");
        free(l);
        return NULL;
    }

    l->stat.hash_bytes = hashsize(l->hashpower);
    return l;
}

static uint32_t 
hash(const char *key, const int nkey)
{
  const char *p;
  int i;
  uint32_t h = 5381;

  for(i = 0, p = key; i < nkey; p++, i++) {
    h = (h << 5) + h + *p;
  }

  return h;
}

static inline lru_item*
do_item_get(lru *l, const char *key, const size_t nkey, const uint32_t hv)
{
    return hash_find(l, key, nkey, hv);
}

static void
do_item_remove_hv(lru *l, lru_item *it, const uint32_t hv) 
{
    assert(it != NULL);
    hash_delete(l, ITEM_key(it), it->nkey, hv);
    if (l->head == it) {
        assert(it->prev == 0);
        l->head = it->next;
    }
    if (l->tail == it) {
        assert(it->next == 0);
        l->tail = it->prev;
    }
    assert(it->next != it);
    assert(it->prev != it);

    if (it->next) it->next->prev = it->prev;
    if (it->prev) it->prev->next = it->next;

    l->stat.curr_items --;
    l->stat.curr_bytes -= it->nbytes;
    l->stat.free_bytes += it->nbytes;
    l->stat.free++;
    free(it);
}

static void
do_item_remove(lru *l, lru_item *it)
{
    uint32_t hv = hash(ITEM_key(it), it->nkey);
    do_item_remove_hv(l, it, hv);
}

int
item_get(lru *l, const char *key, const size_t nkey, char *buf, const size_t nbuf, size_t *nvalue)
{
    l->stat.get_cmds++;
    uint32_t hv = hash(key, nkey);
    lru_item *it = do_item_get(l, key, nkey, hv);
    if (it != NULL) {
        l->stat.get_hits ++;
        size_t vlen = it->nbytes - ITEM_size - it->nkey - 1;

        if (nvalue) {
            *nvalue = vlen;
        }

        if (buf && nbuf) {
            void *data = ITEM_data(it);
            memcpy(buf, data, MIN(nbuf, vlen));
        }
        return 0;
    }
    l->stat.get_misses ++;
    return 1;
}

static void*
item_alloc(lru *l, const size_t sz, lru_item *old)
{
    void *m = NULL;
    int delta = old ? old->nbytes : 0;
    if ((l->stat.curr_bytes + sz - delta) <= l->max_bytes) {
        
        m = malloc(sz);
        if (!m) {
            l->stat.malloc_failed += 1;
            return NULL;
        }

    } else if (sz > l->max_bytes) {
        return NULL;

    } else {
        //evict 
        assert(l->tail != NULL);
        lru_item *it = l->tail;
        while((it != NULL) && (l->stat.curr_bytes + sz - delta) > l->max_bytes) {
            if (it != old) {
                do_item_remove(l, it);
                it = l->tail;
            } else {
                it = l->tail->prev;
            }

            l->stat.evictions ++;
        }
        m = malloc(sz);
        if (!m) {
            l->stat.malloc_failed += 1;
            return NULL;
        }
    }
    l->stat.malloc ++;
    l->stat.curr_bytes += sz;
    l->stat.total_bytes += sz;
    l->stat.curr_items ++;
    l->stat.total_items ++;
    return m;
}

int 
item_set(lru *l, const char *key, const size_t nkey, const char *value, const size_t nvalue)
{
    l->stat.set_cmds ++;

    uint32_t hv = hash(key, nkey);
    lru_item *old = do_item_get(l, key, nkey, hv);

    size_t isize = ITEM_size + nkey + 1 + nvalue;
    lru_item *it = item_alloc(l, isize, old);
    if (it == NULL) {
        l->stat.set_failed ++;
        return 1;
    }

    if (old != NULL) {
        do_item_remove_hv(l, old, hv);
    }

    it->nkey = nkey;
    it->nbytes = isize;

    memcpy(ITEM_key(it), key, nkey);
    memcpy(ITEM_data(it), value, nvalue);
    
    hash_insert(l, it, hv);
    it->prev = 0;
    it->next = l->head;
    if (it->next) it->next->prev = it;
    l->head = it;
    if (l->tail == NULL) l->tail = it;
    return 0;
}

int 
item_delete(lru *l, const char *key, const size_t nkey)
{
    l->stat.del_cmds++;
    uint32_t hv = hash(key, nkey);
    lru_item *it = hash_find(l, key, nkey, hv);

    if (it != NULL) {
        do_item_remove_hv(l, it, hv);
        l->stat.del_hits++;
        return 0;
    } else {
        l->stat.del_misses++;
        return 1;
    }
}

static void 
append_stat(char **buf, int *nbuf, char *name, const char *fmt, ...)
{

    if (*nbuf <= 0) {
        return;
    }

    char val_str[16];
    int len;
    va_list ap;
    va_start(ap, fmt);
    vsnprintf(val_str, sizeof(val_str) - 1, fmt, ap);
    va_end(ap);

    len = snprintf(*buf, *nbuf, "%s = %s\r\n", name, val_str);
    *nbuf -= len;
    *buf += len;
}

void 
stat_print(lru *l, char *buf, const int nbuf)
{
    int remaining = nbuf;

    append_stat(&buf, &remaining, "max_bytes","%llu", l->max_bytes);
    append_stat(&buf, &remaining, "total_items","%llu", l->stat.total_items);
    append_stat(&buf, &remaining, "curr_items","%llu", l->stat.curr_items);
    append_stat(&buf, &remaining, "total_bytes","%llu", l->stat.total_bytes);
    append_stat(&buf, &remaining, "curr_bytes","%llu", l->stat.curr_bytes);
    append_stat(&buf, &remaining, "malloc","%llu", l->stat.malloc);
    append_stat(&buf, &remaining, "malloc_failed","%llu", l->stat.malloc_failed);
    append_stat(&buf, &remaining, "free_bytes","%llu", l->stat.free_bytes);
    append_stat(&buf, &remaining, "free","%llu", l->stat.free);

    append_stat(&buf, &remaining, "get_cmds","%llu", l->stat.get_cmds);
    append_stat(&buf, &remaining, "get_hits","%llu", l->stat.get_hits);
    append_stat(&buf, &remaining, "get_misses","%llu", l->stat.get_misses);

    append_stat(&buf, &remaining, "set_cmds","%llu", l->stat.set_cmds);
    append_stat(&buf, &remaining, "set_failed","%llu", l->stat.set_failed);

    append_stat(&buf, &remaining, "del_cmds","%llu", l->stat.del_cmds);
    append_stat(&buf, &remaining, "del_hits","%llu", l->stat.del_hits);
    append_stat(&buf, &remaining, "del_misses","%llu", l->stat.del_misses);

    append_stat(&buf, &remaining, "hash_power_level","%u", l->hashpower);
    append_stat(&buf, &remaining, "hash_find_depth","%u", l->stat.hash_find_depth);
    append_stat(&buf, &remaining, "hash_bytes","%llu", l->stat.hash_bytes);

    append_stat(&buf, &remaining, "evictions","%llu", l->stat.evictions);
}

void 
lru_free(lru *l)
{
    //free items
    lru_item *it;
    it = l->head;
    while (it != NULL) {
        do_item_remove(l, it);
        it = l->head;
    }
    assert(l->head == NULL);
    assert(l->tail == NULL);
    //free hashtable
    free(l->table);
    free(l);
}
